clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -908,311 +628,781 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
-        Context.getDiagnostics().Report(
-            Fld->getLocation(),
-            diag::err_sycl_non_trivially_copy_ctor_dtor_type)
-            << 1 << ArgTy;
-        AllArgsAreValid = false;
-        continue;
-      }
+    MemberExpr *SpecialObjME = BuildMemberExpr(Base, Field);
+    MemberExpr *MethodME = BuildMemberExpr(SpecialObjME, Method);
+
+    QualType ResultTy = Method->getReturnType();
+    ExprValueKind VK = Expr::getValueKindForType(ResultTy);
+    ResultTy = ResultTy.getNonLValueExprType(SemaRef.Context);
+    llvm::SmallVector<Expr *, 4> ParamStmts;
+    const auto *Proto = cast<FunctionProtoType>(Method->getType());
+    SemaRef.GatherArgumentsForCall(SourceLocation(), Method, Proto, 0,
+                                   ParamDREs, ParamStmts);
+    // [kernel_obj or wrapper object].accessor.__init(_ValueType*,
+    // range<int>, range<int>, id<int>)
+    CXXMemberCallExpr *Call = CXXMemberCallExpr::Create(
+        SemaRef.Context, MethodME, ParamStmts, ResultTy, VK, SourceLocation());
+    if (MethodName == FinalizeMethodName)
+      FinalizeStmts.push_back(Call);
+    else
+      BodyStmts.push_back(Call);
+  }
 
-      CreateAndAddPrmDsc(Fld, ArgTy);
-
-      // Create descriptors for each accessor field in the class or struct
-      createParamDescForWrappedAccessors(Fld, ArgTy);
-    } else if (ArgTy->isReferenceType()) {
-      Context.getDiagnostics().Report(
-          Fld->getLocation(), diag::err_bad_kernel_param_type) << ArgTy;
-      AllArgsAreValid = false;
-    } else if (ArgTy->isPointerType()) {
-      // Pointer Arguments need to be in the global address space
-      QualType PointeeTy = ArgTy->getPointeeType();
-      Qualifiers Quals = PointeeTy.getQualifiers();
-      Quals.setAddressSpace(LangAS::opencl_global);
-      PointeeTy =
-          Context.getQualifiedType(PointeeTy.getUnqualifiedType(), Quals);
-      QualType ModTy = Context.getPointerType(PointeeTy);
-
-      CreateAndAddPrmDsc(Fld, ModTy);
-    } else if (ArgTy->isScalarType()) {
-      CreateAndAddPrmDsc(Fld, ArgTy);
-    } else {
-      llvm_unreachable("Unsupported kernel parameter type");
+  // TODO Remove kernel obj clone
+  static VarDecl *createKernelObjClone(ASTContext &Ctx, DeclContext *DC,
+                                       CXXRecordDecl *KernelObj) {
+    TypeSourceInfo *TSInfo =
+        KernelObj->isLambda() ? KernelObj->getLambdaTypeInfo() : nullptr;
+    VarDecl *VD = VarDecl::Create(
+        Ctx, DC, SourceLocation(), SourceLocation(), KernelObj->getIdentifier(),
+        QualType(KernelObj->getTypeForDecl(), 0), TSInfo, SC_None);
+
+    return VD;
+  }
+
+  void handleSpecialType(FieldDecl *FD, QualType Ty) {
+    const auto *RecordDecl = Ty->getAsCXXRecordDecl();
+    // TODO : we don't need all this init stuff if remove kernel obj clone
+    // Perform initialization only if it is field of kernel object
+    if (MemberExprBases.size() == 1) {
+      InitializedEntity Entity =
```

Comment:
Is this similar enough to the stuff in createExprForStructOrScalar that we could extract a function?

Reply:
No, it does other thing.

Reply:
Ok, it looked like there might be a function to do much of the initializedentity bits that could be shared between the two.